---
title: "230622 PanNET primary vs metastasis"
date: "__`r lubridate::today()`__"
output: 
  html_document:
    number_sections: true
    code_folding: hide
    toc: true
    toc_depth: 1
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = T)

library(tidyverse)
library(openxlsx)
library(ggrepel)
library(kableExtra)
library(DESeq2)

expression_dir = "/Bioinformatics/projects/006_Tumoroids_Viper"
```

# Aim

As preliminary data for the expected differences between PanNET primary tumors an metastases published data in analyzed. 

# Alvarez 2018

In [Alvarez 2018](https://doi.org/10.1038/s41588-018-0138-4) neuroendocrine tumors were analyzed using the viper network activity inference. The data set contains unmatched primary tumors and metastases of pancreatic origin.  

```{r loading counts}
GEP_colData = readRDS(file.path(expression_dir, "data_objects/GepNET_meta_data.Rds")) %>% 
  filter(origin == "pancreas") %>% 
  mutate(type = factor(type, levels = c("primary", "liver.metastasis")))

GEP = readRDS(file.path(expression_dir, "data_objects/GepNET_count_matrix.Rds"))
GEP = GEP[, GEP_colData$sample_name]

GEP_colData %>% 
  group_by(type) %>% 
  dplyr::count() %>% 
  kable() %>% 
  kable_styling(bootstrap_options = "striped", full_width = F, font_size = 10)
```

## All samples 

Differentially expressed genes are identified using DESeq2. This matches the processing applied in Alvarez 2018.  

* minimum count of 10 (adjusted for library size `edgeR::filterByExpr`)

```{r GEP DESeq}
GEP_is_expressed = edgeR::filterByExpr(GEP, 
                                       min.count = 10, 
                                       group = GEP_colData$type)

GEP = GEP[GEP_is_expressed, ]

if (!exists("GEP_DESeq_shrunk")){
  
  GEP_DESeq = DESeqDataSetFromMatrix(countData = GEP, 
                                     colData = GEP_colData, 
                                     design = ~ type)
  GEP_DESeq = DESeq2::estimateSizeFactors(GEP_DESeq)
  GEP_DESeq_res = DESeq(GEP_DESeq)
  GEP_DESeq_shrunk = list(met_vs_prim = lfcShrink(GEP_DESeq_res, 
                                                  coef = "type_liver.metastasis_vs_primary", 
                                                  type = "apeglm"))
  
  # The PCA is based on the vst normalized counts
  GEP_vst = vst(GEP_DESeq, blind = F)
}

```

### PCA 

* vst-normalized (`DESeq`) counts 
* top 2000 most variable genes
* no scaling

```{r GEP pca plot, fig.height = 3.5, fig.width = 5}
# selecting most variable genes 
GEP_mostVar = assay(GEP_vst)
GEP_mostVar = GEP_mostVar[order(rowVars(GEP_mostVar), decreasing = T)[1:2000], ]

GEP_pcaData = prcomp(t(GEP_mostVar), scale. = F)
GEP_percentVar = round(100 * GEP_pcaData$sdev^2/sum(GEP_pcaData$sdev^2), digits = 1)

GEP_pcaData = bind_cols(as.data.frame(GEP_pcaData$x), GEP_colData)

ggplot(GEP_pcaData, 
       aes(x = PC1, 
           y = PC2,
           color = type)) + 
  geom_point(alpha = 0.75) + 
  scale_color_brewer(palette = "Set1") +
  theme_classic() + 
  labs(x = paste0("PC1 (", round(GEP_percentVar[1], digits = 2), "% variance)"),
       y = paste0("PC2 (", round(GEP_percentVar[2], digits = 2), "% variance)"))
```

* Metastases and primary tumors show weak separation along PC1 with few clear outlier samples

### Differentially expressed genes  

* design `~ type`
* shrinkage of log fold changes via `apeglm`

```{r MA plots and pHistograms}
getPHisto = function(DEresult, contrast){
  
  ggplot(filter(as.data.frame(DEresult[[contrast]]), baseMean > 1 ), aes(pvalue)) + 
    geom_histogram(binwidth = 0.05, 
                   color = "white", 
                   na.rm = T, 
                   center = 0.025 ) + 
    theme_classic() + 
    
    labs(x = "unadjusted p-value", title = paste0(contrast,"\ndistribution of p-values")) + 
    
    theme(axis.text = element_text(size = 10, color = "black"), 
          axis.title = element_text(size = 12) ,
          axis.ticks = element_line(color = "black"))
}


getMAPlots = function(DEresult, contrast, ylim = c(-3, 3), sig.threshold = 0.1){
  MA.data = as.data.frame(DEresult[[contrast]])
  MA.data = MA.data[MA.data$baseMean != 0, ]
  MA.data$outside.range = ifelse(MA.data$log2FoldChange > min(ylim), 
                                 ifelse(MA.data$log2FoldChange <= max(ylim), "b", "c"), 
                                 "a")
  MA.data$log2FoldChange[MA.data$outside.range == "a"] = min(ylim)
  MA.data$log2FoldChange[MA.data$outside.range == "c"] = max(ylim)
  MA.data$sig = ifelse(MA.data$pvalue < sig.threshold, T, F)
  MA.data = MA.data %>% arrange(sig)
  
  ggplot(MA.data, aes(x = baseMean, 
                      y = log2FoldChange, 
                      shape = outside.range, 
                      color = sig)) +
    geom_point(alpha = 0.5,
               show.legend = F) + 
    geom_hline(yintercept = 0, size = 2, alpha = 0.5) + 
    scale_color_manual(values = c("grey70", "blue")) +
    scale_shape_manual(values = c(a = 6, b = 20, c = 2)) + 
    scale_x_log10() + 
    scale_y_continuous(breaks = seq(floor(min(ylim)), ceiling(max(ylim))), limits = c(floor(min(ylim)), ceiling(max(ylim)))) + 
    theme_classic() + 
    theme(axis.text = element_text(size = 10, color = "black")) + 
    labs(x = "log10 mean of normalized counts", y = "log fold change", title = paste0(contrast,"\nMA plot")) 
}
```

```{r GEP type result, fig.height = 4, fig.width = 8}
cowplot::plot_grid(
  getMAPlots(GEP_DESeq_shrunk, "met_vs_prim", 
             ylim = c(-6,6),
             sig.threshold = 0.05), 
  getPHisto(GEP_DESeq_shrunk, "met_vs_prim"))

```

* There are `r sum(GEP_DESeq_shrunk$met_vs_prim$padj < 0.05)` genes with an adjusted p values < 0.05 (blue dots)
* Of these, `r sum(GEP_DESeq_shrunk$met_vs_prim$padj < 0.05 & abs(GEP_DESeq_shrunk$met_vs_prim$log2FoldChange) > 1)` also have an absolute log2 fold change > 1

The number of genes with significant differences is relatively large given the overlap between the two conditions. This most likely is due to the outlier metastasis samples. Additionally, the PCA does only capture a small fraction of all the variability between samples. Also the large sample size makes it easier to identify significant differences.    

## Removing outlier samples 

To check for the impact of the three outlier samples (PC1 < -100), they are removed from the data set.  

All other processing parameters remain the same, the removal of lowly expressed genes is not redone.    

```{r GEPfilt DESeq}
if (!exists("GEPfilt_DESeq_shrunk")){
  
  GEPfilt_colData = GEP_colData[GEP_pcaData$PC1 > -100, ]
  GEPfilt = GEP[, GEPfilt_colData$sample_name]
  
  GEPfilt_DESeq = DESeqDataSetFromMatrix(countData = GEPfilt, 
                                         colData = GEPfilt_colData, 
                                         design = ~ type)
  GEPfilt_DESeq = DESeq2::estimateSizeFactors(GEPfilt_DESeq)
  GEPfilt_DESeq_res = DESeq(GEPfilt_DESeq)
  GEPfilt_DESeq_shrunk = list(met_vs_prim = lfcShrink(GEPfilt_DESeq_res, 
                                                      coef = "type_liver.metastasis_vs_primary", 
                                                      type = "apeglm"))
  
  # The PCA is based on the vst normalized counts
  GEPfilt_vst = vst(GEPfilt_DESeq, blind = F)
}
```

## PCA

```{r GEPfilt pca plot, fig.height = 3.5, fig.width = 5}
# selecting most variable genes 
GEPfilt_mostVar = assay(GEPfilt_vst)
GEPfilt_mostVar = GEPfilt_mostVar[order(rowVars(GEPfilt_mostVar), decreasing = T)[1:2000], ]

GEPfilt_pcaData = prcomp(t(GEPfilt_mostVar), scale. = F)
GEPfilt_percentVar = round(100 * GEPfilt_pcaData$sdev^2/sum(GEPfilt_pcaData$sdev^2), digits = 1)

GEPfilt_pcaData = bind_cols(as.data.frame(GEPfilt_pcaData$x), GEPfilt_colData)

ggplot(GEPfilt_pcaData, 
       aes(x = PC1, 
           y = PC2,
           color = type)) + 
  geom_point(alpha = 0.75) + 
  scale_color_brewer(palette = "Set1") +
  theme_classic() + 
  labs(x = paste0("PC1 (", round(GEPfilt_percentVar[1], digits = 2), "% variance)"),
       y = paste0("PC2 (", round(GEPfilt_percentVar[2], digits = 2), "% variance)"))
```

### Differentially expressed genes  

```{r GEPfilt type result, fig.height = 4, fig.width = 8}
cowplot::plot_grid(
  getMAPlots(GEPfilt_DESeq_shrunk, "met_vs_prim", 
             ylim = c(-6,6),
             sig.threshold = 0.05), 
  getPHisto(GEPfilt_DESeq_shrunk, "met_vs_prim"))

```

* There are `r sum(GEPfilt_DESeq_shrunk$met_vs_prim$padj < 0.05)` genes with an adjusted p values < 0.05 (blue dots)
* Of these, `r sum(GEPfilt_DESeq_shrunk$met_vs_prim$padj < 0.05 & abs(GEPfilt_DESeq_shrunk$met_vs_prim$log2FoldChange) > 1)` also have an absolute log2 fold change > 1

While the number of differentially expressed genes is reduced when outliers are excluded, a substantial number of hits does remain. The PCA explains only a small fraction of the total variance and does not fully illustrate the differnces between the groups. Also the large sample size makes it easier to identify significant differences.   

### Comparison of results  

The p values from the test including or excluding metastases are compared.

```{r GEP compare p values, fig.height = 4, fig.width = 6}
stopifnot(all.equal(rownames(GEP_DESeq_shrunk$met_vs_prim),
                    rownames(GEPfilt_DESeq_shrunk$met_vs_prim)))

data.frame(pval_outlier = GEP_DESeq_shrunk$met_vs_prim$pvalue,
           padj_outlier = GEP_DESeq_shrunk$met_vs_prim$padj,
           pval_no_outlier = GEPfilt_DESeq_shrunk$met_vs_prim$pvalue,
           padj_no_outlier = GEPfilt_DESeq_shrunk$met_vs_prim$padj) %>% 
  mutate(sig_in_both = ifelse(padj_outlier < 0.05 & padj_no_outlier < 0.05, "yes", "no")) %>% 
  ggplot(aes(x = -log10(pval_outlier),
             y = -log10(pval_no_outlier),
             color = sig_in_both)) + 
  geom_point(alpha = 0.5) + 
  geom_hline(yintercept = -log10(0.05),
             linetype = "dashed") + 
  geom_vline(xintercept = -log10(0.05),
             linetype = "dashed") +
  geom_abline(slope = 1, intercept = 0, 
              linetype = "dotted") +
  scale_color_manual(values = c(yes = "blue", no = "grey60")) + 
  theme_classic() + 
  labs(x = "-log10 (p value) with outlier",
       y = "-log10 (p value) without outlier",
       color = "adjusted pavalue < 0.05\nin both tests")
```

* All of the top hits are conserved between the two tests
* Many p values are along the dotted line indicating equality
* A group of genes gets much smaler p values when the outliers are included

Based on this comparison, it becomes clear that the analysis with outliers includes many genes with comparatively large p values. Therefore, the data set without the outliers is preferred.  

# Summary Alvarez 

The results for both analyses are saved under `230622_PanNET_primary_vs_metastasis.xlsx` on separate worksheets.  

Due to the log2 fold change shrinkage p values of highly variable genes may be set to NA. These need to be considered as non-significant.  

The published genes lists are based on Entrez gene IDs. To convert the IDs to gene names the Bioconductor 3.16 version of `org.Hs.eg.db` is used. This is not the version matched to the original analysis (not specified in the methods section of the paper).  

```{r saving results, eval = F}
gene_anno = AnnotationDbi::select(org.Hs.eg.db, 
                             keys = rownames(GEP), 
                             keytype = "ENTREZID", 
                             column = c("SYMBOL", "GENENAME", "GENETYPE")) %>% 
  dplyr::rename(Entrez_id = ENTREZID,
                gene_name = SYMBOL,
                gene_description = GENENAME,
                gene_type = GENETYPE)

annotateResults = function(DESeq_res){
  DESeq_res = DESeq_res %>% 
    as.data.frame() %>% 
    rownames_to_column("Entrez_id") 
  
  left_join(gene_anno,
            DESeq_res,
            by = "Entrez_id") %>% 
    arrange(padj)
}

write.xlsx(list(met_vs_prim_with_outlier = annotateResults(GEP_DESeq_shrunk$met_vs_prim),
                met_vs_prim_without_outlier = annotateResults(GEPfilt_DESeq_shrunk$met_vs_prim)),
           "230622_PanNET_primary_vs_metastasis.xlsx",
           overwrite = T)

```
```{r session info}
sessionInfo()
```

------

