---
title: "PAnNET QA Analysis Sample Data"
date: "__`r lubridate::today()`__"
output: 
  html_document:
    number_sections: true
    code_folding: hide
    toc: true
    toc_depth: 1
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)

library(DESeq2)
library(readr)
library(RColorBrewer)
library(ggplot2)
library(gplots)
library(pheatmap)
library(ggrepel)
library("BiocParallel")
library(ggpubr)
library(factoextra)
library("PoiClaClu")
library(ggrepel)
library(cowplot)
library(readxl)
library(tidyverse)
library(dplyr)
library(tidyr)
library(stringr)
library(DT)

library(foreach)
library(doParallel)

register(MulticoreParam(4))

expression_dir = "C:/Users/br23n469/Desktop/PanNet_RNASeq_Analysis"

```

# Aim

The main goal of this study is to analyze the data thoroughly using DESeq2 before applying PCA, heatmap, and distance matrix. We use DESeq2 to assess the subread count, as well as interactively query and explore the dataset, in an attempt to find genes that are differentially expressed. By applying PCA, distance matrix plots and K-means Clustering, the structure and interaction of the data may be further highlighted, making it simpler to identify clusters, similarities, and variations.

# Analyses of Data

## Loading the Data

* Sub-read count Data is loaded and also removed columns which are not need. This columns which were removed:
  + Chr 
  + Start 
  + End 
  + Strand 
  + Length

```{r Loading data, echo=TRUE}

# Setting up color profiles from color brewer
qual_col_pals = brewer.pal.info[brewer.pal.info$category == 'qual',]
mycols = unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))

# Sample names
sample_1 <- "T.1"
sample_2 <- "T.2"
sample_3 <- "T.3"
sample_4 <- "T.4"

# Create Samples Data-frame
condition_column <- c(rep(sample_1, 8),
                      rep(sample_2, 8),
                      rep(sample_3, 6),
                      rep(sample_4, 3))

run_column <- paste0(condition_column,
                     ".",
                     c(1:8, 1:8, 1:6, 1:3))

rep_column <- gsub("T.", "", run_column)

samples_df <- data.frame(condition = condition_column,
                         run = run_column,
                         rep = rep_column, 
                         row.names = run_column)

# Convert 'condition' column to factor
samples_df$condition = factor(samples_df$condition)

# The table contains values for each CDS included in the sub_read_count_data variable.
sub_read_count_data <- read.table("subreadcounts.txt", header = TRUE, row.names = 1)


# Remove specific columns from sub_read_count_data like 'Chr', 'Start', 'End', 'Strand' and 'Length'.
sub_read_count_data <- sub_read_count_data[ , -c(3, 4, 1, 2,5)]

# Change column names
colnames(sub_read_count_data) <- rownames(samples_df)

```

* Adding 'Tissue for co_extraction_DESI_DNA_RNA_rep' data. 
* First 25 rows were extracted.
* Columns of interest includes: 
  + Date of isolation
  + Id
  + B-Nr_Original
  + B-Nr.
  + RQN from Vetsuisse
  + Vetsuisse results after their Qubit

```{r processing co_extraction DESI DNA RNA data, echo=TRUE}

# Read the DESI_DNA_RNA_rep Excel file, sheet name = Total RNA extraction.
desiDNARNARep <- suppressMessages(read_excel("meta_data/20020505_Tissue for co_extraction_DESI_DNA_RNA_rep.update_23.9.22.xlsx", sheet = "Total RNA extraction", 
                                             .name_repair = "unique"))

# Specify the column names you want to extract
vectorNameColumns <- c("Date of isolation", "...3", "B-Nr_Original", "B-Nr.", "RQN from Vetsuisse", 
                       "Vetsuisse results after their Qubit")

# Specify the row range you want to extract
vectorNameRows <- 1:25

# Extract the desired rows and columns from the Excel data and convert it to a data frame
df.desiDNARNARep <- as.data.frame(desiDNARNARep[vectorNameRows, vectorNameColumns])

# Rename the column '...3' to 'Id'
colnames(df.desiDNARNARep)[colnames(df.desiDNARNARep) == "...3"] <- "Id"

# Function to replace data starting with dot R. with an empty value
replace_R_with_empty <- function(data) {
  data <- lapply(data, function(col) {
    gsub("^R.", "", col) # Replace data starting with R. with an empyty value
  })
}

df.desiDNARNARep$Id <- (replace_R_with_empty(df.desiDNARNARep$Id))

# Convert data types of the dataframe
df.desiDNARNARep <- type.convert(df.desiDNARNARep, as.is = TRUE)

# Apply the 'str_trim' function to remove leading and trailing whitespace from all columns of 'df.desiDNARNARep'
df.desiDNARNARep <- df.desiDNARNARep %>%
  mutate_all(str_trim)

```

* Adding 'samples for DESI and RNA extraction' data. 
* First 31 rows were extracted.
* Columns of interest includes: 
  + B-Nr.
  + Archivia
  + Geschlecht
  + Organ
  + Reubi /TBBCollection
  + WHO 2017
  + WHO 2017 code
  + CA9 primary tumors
  + MCT1_TMA
  + MCT4_TMA
  + MCT4 (new scoring)
  + DESI
  
```{r sample DESI and RNA extraction data, echo=TRUE}

# Read the samples for DESI and RNA extraction Excel file.
sampleDNARNAExtra <- suppressMessages(read_excel("meta_data/samples for DESI and RNA extraction.xlsx", 
                            sheet = "Sheet2", .name_repair = "unique"))

# Specify the column names you want to extract
vectorNameColumns2 <- c("B-Nr.", "Archivia", "Geschlecht", "Organ",  "Reubi /TBBCollection", 
                       "WHO 2017", "WHO 2017 code", "CA9 primary tumors", 
                       "MCT1_TMA", "MCT4_TMA", "MCT4 (new scoring)", "DESI")

# Specify the row range you want to extract
vectorNameRows2 <- 1:31

# Extract the desired rows and columns from the Excel data and convert it to a data frame
df.sampleDNARNAExtra <- as.data.frame(sampleDNARNAExtra[vectorNameRows2, vectorNameColumns2])

# Convert data types of the dataframe
df.sampleDNARNAExtra <- type.convert(df.sampleDNARNAExtra, as.is = TRUE)

# Apply the 'str_trim' function to remove leading and trailing whitespace 
# from all columns of 'df.sampleDNARNAExtra'
df.sampleDNARNAExtra <- df.sampleDNARNAExtra %>%
  mutate_all(str_trim)
df.sampleDNARNAExtra$Geschlecht <- ifelse(df.sampleDNARNAExtra$Geschlecht == "m", "male", ifelse(df.sampleDNARNAExtra$Geschlecht == "f", "female", "female"))

``` 

* Left joining on the two data frames using the "B-Nr." column.

```{r Joining co_extraction and sample data, echo=TRUE}

# Perform a left join on the two data frames using the "B-Nr." column.
df.desiDNARNASample <- left_join(df.desiDNARNARep, df.sampleDNARNAExtra, by="B-Nr.")

# Add a new "name" by appending "T." to the existing "Id" column.
df.desiDNARNASample$name <- paste0("T.", df.desiDNARNASample$Id)

```

## Meta data

The meta data is extracted from the 'samples for DESI and RNA extraction' file names, with the unique identifiers.

```{r Creating CEL_seq_info}

# Sample information
CEL_seq_info <- df.desiDNARNASample

```

```{r init reads}
# The number of reads is determined using the fastqc output and added to the metadata
init_reads <- read.csv("processed_data/QC/init_fastqc_221027.csv", 
                      header = T, 
                      strip.white = T, 
                      stringsAsFactors = F)

init_reads <- init_reads %>% 
  mutate(name = gsub("_L[0-9].*", "", sample),
         Sample_UID = paste0(name, "-", run)) %>% 
  filter(grepl("R1.*fastq.gz", sample)) %>% 
  group_by(Sample_UID, name) %>% 
  summarise(raw_reads = sum(reads), .groups="drop_last") %>% 
  mutate(name = gsub("_", ".", name))
  
CEL_seq_info <- left_join(CEL_seq_info, init_reads, by = "name",
                           suffix = c(".old", "")) %>%
  dplyr::select(!ends_with("old"))


```

```{r trim reads}
# The reads are processed through read trimming This data is read from the cutadapt logs and saved in a summary file
# For each sequencing run cutadapt is executed separately and the summary is stored in the respective run folder

cutadapt_files <- list.files("processed_data", pattern = "cutadapt.*summary", recursive = T)


trim_reads <- lapply(cutadapt_files, 
                    function(x) read.csv(file.path("processed_data", x)))

names(trim_reads) <- str_extract(cutadapt_files, 
                                pattern = "A[^/]*_[0-9]+")

trim_reads <- bind_rows(trim_reads, 
                       .id = "run")

trim_reads <- trim_reads %>% 
  mutate(name = gsub("_L[0-9].*", "", file),
         Sample_UID = paste0(name, "-", run)) %>% 
  dplyr::rename(cutadapt_in = reads_in,
                cutadapt_out = reads_out) %>% 
  group_by(Sample_UID, name) %>%
  summarise(across(where(is.integer),  ~ sum(.x)), .groups="drop_last") %>%
  mutate(cutadapt_out_perc = round(100*(cutadapt_out / cutadapt_in), digits = 2),
         name = gsub("_", ".", name))

CEL_seq_info <- left_join(CEL_seq_info, trim_reads, 
                          by=c('Sample_UID'='Sample_UID', 'name'='name'),
                          suffix = c(".old", ""))  %>% 
  dplyr::select(!ends_with("old"))  %>% 
  mutate(cutadapt_out_perc = round(100*(cutadapt_out / cutadapt_in), 
                                   digits = 2))
  
```

```{r ribodetector}
# Ribosomal reads are removed using ribodetector
# The remaining reads can be found in the fastqc output after rRNA removal
filt_reads <- read.csv("processed_data/QC/filt_fastqc_230717.csv", 
                      header = T, 
                      strip.white = T, 
                      stringsAsFactors = F)

filt_reads<- filt_reads %>% 
  mutate(name = gsub("_L[0-9].*", "", sample),
         Sample_UID = paste0(name, "-", run)) %>% 
  filter(grepl("R1.*fastq.gz", sample)) %>% 
  group_by(Sample_UID, name) %>% 
  summarise(filt_reads = sum(reads), .groups="drop_last") %>% 
  mutate(name = gsub("_", ".", name))

CEL_seq_info <- left_join(CEL_seq_info, filt_reads, 
                          by=c('Sample_UID'='Sample_UID', 'name'='name'), 
                          suffix = c(".old", "")) %>% 
  dplyr::select(!ends_with("old")) %>% 
  mutate(filt_reads_perc = round(100*(filt_reads / cutadapt_out), digits = 2))
  

```

```{r mapped reads}
STAR_files <- list.files("processed_data", pattern = "STAR.*mapped_reads.txt", recursive = T)

STAR_reads<- lapply(STAR_files, 
                    function(x) read.csv(file.path("processed_data", x)))
names(STAR_reads) <- str_extract(STAR_files, 
                                pattern = "A[^/]*_[0-9]+")
STAR_reads <- bind_rows(STAR_reads, 
                       .id = "run")

STAR_reads <- STAR_reads %>% 
  mutate(name = gsub("_L[0-9].*", "", file),
         Sample_UID = paste0(name, "-", run)) %>% 
  dplyr::rename(STAR_out = mapped_reads) %>% 
  group_by(Sample_UID, name) %>% 
  summarise(across(where(is.integer),  ~ sum(.x)), .groups="drop_last") %>% 
  mutate(name = gsub("_", ".", name))

CEL_seq_info <- left_join(CEL_seq_info, STAR_reads, 
                           by=c('Sample_UID'='Sample_UID', 'name'='name'), suffix = c(".old", "")) %>% 
  dplyr::select(!ends_with("old")) %>% 
  mutate(STAR_out_perc = round(100*(STAR_out / filt_reads), digits = 2))

```

```{r reads after subread dedup}
count_reads <- read.delim("processed_data/results_reverse/subreadcounts.txt.summary", 
                         sep = "\t", 
                         strip.white = T, 
                         header = T)
rownames(count_reads) <- count_reads[[1]]
count_reads <- as.data.frame(t(count_reads[, -1]))

count_reads <- count_reads %>% 
  rownames_to_column("name") %>% 
  mutate(name = gsub("\\.bam", "",name),
         name = gsub("_", ".", name)) %>% 
  dplyr::select(name, 
                assigned = Assigned, 
                unassigned_multi = Unassigned_MultiMapping,
                unassigned_nofeatures = Unassigned_NoFeatures, 
                unassigned_ambiguity = Unassigned_Ambiguity) %>% 
  mutate(across(-1, ~ as.integer(.x)))

CEL_seq_info <- left_join(CEL_seq_info, count_reads, 
                           by = "name", suffix = c(".old", "")) %>% 
  dplyr::select(!ends_with("old")) %>% 
  mutate(assigned_perc = round(100*(assigned / STAR_out), digits = 2))

```

```{r add meta data}
# Adding the metadata 
CEL_metaData <- CEL_seq_info

# The meta data is saved to be useful later on
saveRDS(CEL_metaData, 
        "meta_data/CEL_metaData.Rds")
```

```{r processing the raw counts}
CEL_countTable <- read_delim("processed_data/results_reverse/subreadcounts.txt", 
                             skip = 1,
                             show_col_types = FALSE) %>% 
  setNames(gsub("_", ".", names(.))) %>% 
  dplyr::rename(gene_id = Geneid)  

  
# removing the version numbers from the gene ids
CEL_countTable$gene_id = gsub("\\.[0-9]*$", "" , CEL_countTable$gene_id)

makeCountData = function(IDs, countTable, name_suffix){
  countTable = as.data.frame(countTable)
  colnames(countTable) = gsub(name_suffix, "", colnames(countTable))
  rownames(countTable) = countTable$gene_id
  countTable[, IDs]
  
}

CEL_countData = makeCountData(CEL_metaData$name, 
                              countTable = CEL_countTable,
                              name_suffix = "\\.bam")
stopifnot(all.equal(colnames(CEL_countData), CEL_metaData$name))

# The expression of certain gene groups can be used for quality control
getGenePerc = function(colData, 
                       countData, 
                       goi, 
                       goi_name){
  colData[goi_name] = colSums(countData[rownames(countData) %in% goi, ])
  colData[paste0(goi_name, "_perc")] = round((colData[[goi_name]] / colData[["assigned"]]) * 100, 
                                             digits = 2)
  return(colData)
}

# reads mapping to mitochondrial genes 
CEL_metaData = getGenePerc(CEL_metaData, 
                           CEL_countData, 
                           goi = CEL_countTable$gene_id[CEL_countTable$Chr == "chrM"], 
                           goi_name = "mito")
# reads mapping to MALAT1 (extremely high coverage in some samples)
CEL_metaData = getGenePerc(CEL_metaData, 
                           CEL_countData, 
                           goi = "ENSG00000251562", 
                           goi_name = "MALAT1")
# reads mapping to the spike ins 
CEL_metaData = getGenePerc(CEL_metaData, 
                           CEL_countData, 
                           goi = CEL_countTable$gene_id[grepl("^ERCC|^SIR", CEL_countTable$gene_id)], 
                           goi_name = "spikein")

# It may be possible to define detected genes on fpkm but this is not really necessary 
CEL_metaData$detected_genes = colSums(CEL_countData > 10)
CEL_metaData$filtered_reads = colSums(CEL_countData)

```

## Data quality assessment

### Overview over the processing

* library preparation `TruSeq stranded total RNA` 
* adapter trimming and removal of long homopolymers (`cutadapt`)
* removal of ribosomal sequences (`ribodetector`)
* mapping to GRCh38, Ensembl v104 (`STAR`)
* assigning reads to genes (`subread featurecounts`)

```{r sample information unitil rRNA}
CEL_metaData %>% 
  dplyr::select(c(name, raw_reads, 
                  cutadapt_out:cutadapt_out_perc, 
                  filt_reads, filt_reads_perc)) %>%
  rename("sample name" = name) %>%
  datatable(options = list(
    initComplete = JS(
      "function(settings, json) {",
      "$(this.api().table().header()).css({'font-size': '10px'});",
      "$(this.api().table().body()).css({'font-size': '9px'});",
      "}"),
    columnDefs = list(list(className = 'dt-center', targets = "_all"))
  ),
  class = 'cell-border stripe table-condensed',
  rownames = F,
  colnames = gsub("_", "\n", colnames(.)),
  filter = "top",
  caption = "RNA-Seq processing from raw reads to ribosomal contamination")


```

* For each sample about 40 million reads were sequenced  
* Adapter content is low with 2%
* The adapter content was higher in the R2 compared to the R1; In most cases the removed sequences are long stretches of the same base (mainly poly-G)
* Ribosomal RNAs are removed using the `ribodetector` tool; This detects about 5% of all sequences as ribosomal.


```{r sample information from mapping}
CEL_metaData %>% 
  dplyr::select(c(name, filt_reads, 
                  STAR_out:assigned_perc)) %>% 
  rename("sample name" = name) %>%
  datatable(options = list(
    initComplete = JS(
      "function(settings, json) {",
      "$(this.api().table().header()).css({'font-size': '10px'});",
      "$(this.api().table().body()).css({'font-size': '9px'});",
      "}"),
    columnDefs = list(list(className = 'dt-center', targets = "_all"))
  ),
  class = 'cell-border stripe table-condensed',
  rownames = F,
  colnames = gsub("_", "\n", colnames(.)),
  filter = "top",
  caption = "RNA-Seq processing from mapping to gene counting") 
```


* Approximately 87% to 92% of read pairs can be mapped uniquely.
* Surprisingly, only between 40% and 70% of reads can be assigned to genomic features.

### Differentially expressed genes with DESeq2

The geometric mean of all genes was determined inside DESeq2 to achieve an internal normalization. This average is then used to normalize gene counts across samples[2].

The DESeq2 package requires count data in the form of an integer matrix, such as that collected from an RNA-seq, as its input. DESeq2 eliminates outliers using Cook's distance. It also eliminates genes whose normalized mean counts are below a threshold. Removing low-count genes enhances detection power by reducing repeated p-value adjustments[1][2].

<!-- At this point we do not need to run DESeq yet (saves computation time) --> 

```{r set up DESeq, include = T, echo=TRUE}

# Import as DESeqDataSet(dds)
if (!exists("dds")){
  dds <- DESeqDataSetFromMatrix(countData = sub_read_count_data,
                                colData = samples_df,
                                design = ~ 1)
  
  # Pre-filtering
  keep <- rowSums(counts(dds)) >= 10
  dds <- dds[keep, ]
}

# Differential expression analysis
# dds <- DESeq(dds)

```

* This creates a DESeqDataSet object from count data and sample metadata.
* It performs pre-filtering by retaining only the genes with a count greater than or equal to 10.
* After filtering `r sum(keep)` genes remaining.

### Normalizing Data

* Any data analysis must include the stages of data quality assessment and quality control.
* Here we have used the 'vst' transformation functions which works must faster when compared to 'rlog' function.

```{r normalizing data, echo=TRUE}

# Variance stabilizing transformation for data quality assessment
vsd <- vst(dds, blind = FALSE)

```

### K-means

* Data - vst-normalized (`DESeq`) counts.
* Number of cluster - 2
* Top 500 most variable genes.
# No scaling was done to data.

```{r K-Means,  fig.height = 4, fig.width = 5, echo=TRUE}

# Compute k-means
set.seed(123)
         
# Calculate the row variances of the `vsd` object
rv <- rowVars(assay(vsd))

# Select the top 500 rows with the highest variances.
select <- order(rv, decreasing = TRUE)[seq_len(min(500, length(rv)))]

# Subset the `vsd` object to include only the selected rows and transpose the result
subset_vsd <- vsd[select, ]

# Perform k-means clustering on the transposed subset of the `vsd` matrix
res.km <- kmeans(t(assay(subset_vsd)), 2, nstart = 20)

# ploting k-means
plotKmeans <- function(data) {
  
  # Dimension reduction using PCA
  res.pca <- prcomp(t(assay(data)))
  
  # Coordinates of individuals
  ind.coord <- as.data.frame(get_pca_ind(res.pca)$coord)
  
  # Add clusters obtained using the K-means algorithm
  ind.coord$cluster <- factor(res.km$cluster)
  
  # Add condition and row.names from the metadata
  ind.coord$condition <- data$condition
  ind.coord$row.names <- colnames(data)
  
  
  # Percentage of variance explained by dimensions
  eigenvalue <- round(get_eigenvalue(res.pca), 1)
  variance.percent <- eigenvalue$variance.percent
  
  # Generate the K-means plot using ggscatter
  ggscatter(
    ind.coord, x = "Dim.1", y = "Dim.2", 
    color = "cluster", palette = "npg", ellipse = TRUE, ellipse.type = "convex",
    shape = "condition",
    size = 3,  legend = "right", ggtheme = theme_bw(),
    xlab = paste0("Dim 1 (", variance.percent[1], "% )" ),
    ylab = paste0("Dim 2 (", variance.percent[2], "% )" ),
    label = "row.names", repel=TRUE, font.label = c(9, "plain","black"),
    title = 'K-means Clustering '
  )
  
}

```


### PCA Plots

* Data - vst-normalized (`DESeq`) counts.
* No scaling was done to data.
* Top 500 most variable genes were taken from a list of 36010 genes.

```{r pcaPlot,  fig.height = 3.25, fig.width = 8, echo=TRUE}

# Perform PCA on the `vsd` data with grouping by "condition" and "rep".
pcaData <- plotPCA(vsd, intgroup = c("condition", "rep"), returnData = TRUE)

# Retrieve the percentage of variance explained by each principal component
percentVar <- round(100 * attr(pcaData, "percentVar"))

pcaPlot <- ggplot(pcaData, aes(PC1, PC2, color = condition, shape = condition)) + 
  geom_point(size = 3) + geom_text_repel(aes(label = name), size = 3, color = "black") +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  theme_classic()
 
# Add clusters obtained using the K-means algorithm to pcaData
pcaData$KmeanCluster <- factor(res.km$cluster)
 
# Create a scatter plot of PC1 vs PC2
kmeansPCAPlot <- ggplot(pcaData, aes(PC1, PC2, color = KmeanCluster, shape = condition)) + geom_point(size = 3) + 
  geom_text_repel(aes(label = name), size = 3, color = "black") +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  theme_classic()
 

# Creating PCA plots
plot_grid(pcaPlot, kmeansPCAPlot, ncol=2)

```

### PCA Plots with sample Data

```{r left join pcaData with df.desiDNARNASample,  fig.height = 3.25, fig.width = 8, echo=TRUE}

# Perform a left join on the two data frames using the "B-Nr." column.
pcaData <- left_join(pcaData, df.desiDNARNASample, by="name")

```


```{r pcaPlots with sample data,  fig.height = 6, fig.width = 8, echo=TRUE}
 
names(pcaData) <- gsub(" ", "_", names(pcaData))
 
# Create a scatter plot of PC1 vs PC2
coollectionPCAPlot <- ggplot(pcaData, aes(PC1, PC2, color = pcaData$`Reubi_/TBBCollection`)) + geom_point(size = 3) + 
  geom_text_repel(aes(label = pcaData$rep), size = 3, color = "black") +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  theme_classic() + labs(color = "Collection")

RQNPCAPlot <- ggplot(pcaData, aes(PC1, PC2, color = pcaData$Geschlecht)) + geom_point(size = 3) + 
  geom_text_repel(aes(label = pcaData$rep), size = 3, color = "black") +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  theme_classic() + labs(color = "Gender")

whoPCAPlot <- ggplot(pcaData, aes(PC1, PC2, color = pcaData$WHO_2017)) + geom_point(size = 3) + 
  geom_text_repel(aes(label = pcaData$rep), size = 3, color = "black") +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  theme_classic() + labs(color = "WHO 2017")

ca9PCAPlot <- ggplot(pcaData, aes(PC1, PC2, color = pcaData$CA9_primary_tumors)) + geom_point(size = 3) + 
  geom_text_repel(aes(label = pcaData$rep), size = 3, color = "black") +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  theme_classic() + labs(color = "CA9\nprimary\ntumors ")
 

# Creating PCA plots
plot_grid(coollectionPCAPlot, RQNPCAPlot,whoPCAPlot,ca9PCAPlot, ncol=2, nrow=2, align="h")

```

```{r pcaPlots with mct,  fig.height = 6, fig.width = 8, echo=TRUE}
 
# Create a scatter plot of PC1 vs PC2
MCT1PCAPlot <- ggplot(pcaData, aes(PC1, PC2, color = pcaData$MCT1_TMA)) + geom_point(size = 3) + 
  geom_text_repel(aes(label = pcaData$rep), size = 3, color = "black") +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  theme_classic() + labs(color = "MCT1 TMA")

MCT2PCAPlot <- ggplot(pcaData, aes(PC1, PC2, color = pcaData$MCT4_TMA)) + geom_point(size = 3) + 
  geom_text_repel(aes(label = pcaData$rep), size = 3, color = "black") +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  theme_classic() + labs(color = "MCT4 TMA")

MCT2NewPCAPlot <- ggplot(pcaData, aes(PC1, PC2, color = pcaData$`MCT4_(new_scoring)`)) + geom_point(size = 3) + 
  geom_text_repel(aes(label = pcaData$rep), size = 3, color = "black") +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  theme_classic() + labs(color = "MCT4\n(new_score)")

pcaData$RQN_from_Vetsuisse <- as.numeric(pcaData$RQN_from_Vetsuisse)

RQNPCAPlot <- ggplot(pcaData, aes(PC1, PC2, color = RQN_from_Vetsuisse)) + geom_point(size = 3) + 
  geom_text_repel(aes(label = pcaData$rep), size = 3, color = "black") +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  theme_classic() + labs(color = "RQN from\nVetsuisse")

# Creating PCA plots
plot_grid(MCT1PCAPlot, MCT2PCAPlot,MCT2NewPCAPlot, RQNPCAPlot, ncol=2,  nrow=2, 
          align="h")

```
```{r pcaPlots with assigned_perc,  fig.height = 6, fig.width = 8, echo=TRUE}

# Add clusters obtained using the K-means algorithm to pcaData
pcaData$assigned_perc <- as.numeric(CEL_metaData$assigned_perc)
 
# Create a scatter plot of PC1 vs PC2
assignedPercPCAPlot <- ggplot(pcaData, aes(PC1, PC2, color = pcaData$assigned_perc, shape = condition)) + geom_point(size = 3) + 
  geom_text_repel(aes(label = pcaData$rep), size = 3, color = "black") +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  theme_classic() + labs(color = "Assigned\ngenomic\nfeatures%")

# Calculate the median of assigned_perc
median_assigned_perc <- median(pcaData$assigned_perc)

# Create a new column to indicate whether each row is above or below the median
pcaData$above_median <- ifelse(pcaData$assigned_perc > median_assigned_perc, "Above", "Below")

# Create a scatter plot of PC1 vs PC2 with points colored by the above_median column
assignedPercMedianPlot <- ggplot(pcaData, aes(PC1, PC2, color = above_median, shape = condition)) + 
  geom_point(size = 3) + 
  geom_text_repel(aes(label = rep), size = 3, color = "black") +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) + 
  theme_classic() + 
  labs(color = "Median\nAssigned\nfeatures%") 


# Create a box plot of assigned_perc with different groups
assignePercboxPlot <- ggplot(pcaData, aes(x = factor(condition), y = assigned_perc)) +
  geom_boxplot(fill = "lightblue", color = "black", outlier.shape = NA, utlier.colour = "red",
               outlier.size = 3) +
  geom_jitter(
    color = "darkblue",
    alpha = 0.6,
    width = 0.2
  ) +
  xlab("Condition") +
  ylab("Assigned Genomic Features %") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))  # Rotate x-axis labels for better readability

# Creating PCA plots
plot_grid(assignedPercPCAPlot, assignedPercMedianPlot, assignePercboxPlot, ncol=2,  nrow=2, align="h")

```

* We do not observe any cluster formation when samples were collected from two different collection points.
* Out of the 25 samples taken, `r sum(pcaData$Geschlecht == "male")` were male, and `r sum(pcaData$Geschlecht == "female")` were female. While we observe that more males are clustered together on the left side of the PCA plot, we cannot be certain if    it forms two separate clusters due to the sample size.
* Out of the 25 samples taken, `r sum(pcaData$WHO_2017 == "NET G1")` were classified as Grade 1 (low-grade tumor), which indicates that these cells divide at a low rate and therefore grow slowly. Additionally, `r sum(pcaData$WHO_2017 == "NET G2")` samples were classified as Grade 2 (intermediate-grade tumor), implying that these cells divide at an intermediate rate.

### Heatmap of sample-to-sample distances

* Data - vst-normalized (`DESeq`) counts.
* Number of cluster - 2
* Euclidean distance between samples were calculated.
* Hierarchical Clustering on the distance matrix was done the ward.D2 method

```{r Heatmap of sample-to-sample distances, fig.height = 6, fig.width = 8, echo=TRUE}

plotHmapDistance <- function(norm_data, n_clust) {
  
  # Calculate the distance matrix between samples
  sampleDists <- dist(t(assay(norm_data)))

  # Perform hierarchical clustering on the distance matrix using the ward.D2 method
  sampleGroups <- cutree(hclust(sampleDists, method = "ward.D2"), n_clust)

  # Create an annotation data frame with cluster assignments
  anno_df <- data.frame(clust = as.character(sampleGroups), row.names = colnames(dds))

  # Generate the heatmap plot using pheatmap
  pheatmap(as.matrix(sampleDists),
           col = colorpanel(100, "#2b8cbe", "#7bccc4", "white"),
           clustering_distance_rows = sampleDists,
           clustering_distance_cols = sampleDists,
           clustering_method = "ward.D2",
           annotation_col = anno_df,
           main = "Heatmap using the variance stabilizing transformed values.")
}

plotHmapDistance(vsd, 2)
```

### Heatmap of sample-to-sample distances using Poisson Distance

* Data - DESeq data count.
* Number of cluster - 2
* Poisson distance between samples were calculated.
* Hierarchical Clustering on the distance matrix was done the ward.D2 method

```{r Heatmap of sample-to-sample distances using Poisson Distance, fig.height = 6, fig.width = 8, echo=TRUE}

plotHmapPoissonDistance <- function(data, n_clust) {
  
  # Calculate the distance matrix between samples using Poisson Distance
  samplePoisDist <- PoissonDistance(t(counts(dds)))
  # Convert the distance matrix to a matrix format
  samplePoisDistMatrix <- as.matrix(samplePoisDist$dd)  
  
  # Set row and column names of the matrix as column names of 'dds'
  rownames(samplePoisDistMatrix) <- paste(colnames(dds))  
  colnames(samplePoisDistMatrix) <- paste(colnames(dds))


  # Perform hierarchical clustering on the distance matrix using the ward.D2 method
  sampleGroups <- cutree(hclust(samplePoisDist$dd, method = "ward.D2"), 2)

  # Create an annotation data frame with cluster assignments
  anno_df <- data.frame(clust = as.character(sampleGroups), row.names = colnames(dds))

  # Generate the heatmap plot using pheatmap
  pheatmap(samplePoisDistMatrix,
         clustering_distance_rows = samplePoisDist$dd,
         clustering_distance_cols = samplePoisDist$dd,
         col = colorpanel(100, "#2b8cbe", "#7bccc4", "white"),
         annotation_col = anno_df,
         main = "Heatmap of sample-to-sample distances using the Poisson Distance.")
}

plotHmapPoissonDistance(dds, 2)
```

### Heatmap of sample-to-sample Correlation

* Data - vst-normalized (`DESeq`) counts.
* Number of cluster - 2
* Correlation between samples were calculated.
* Hierarchical Clustering on the distance matrix was done the ward.D2 method

```{r Heatmap of sample-to-sample Correlation, fig.height = 6, fig.width = 8, echo=TRUE}

plotHmapCorrelation <- function (norm_data, n_clust){
  
  # Calculate the correlation matrix of the variance stabilizing transformed data
  sampleCor <- cor(assay(norm_data))
  
  # Perform hierarchical clustering on the correlation matrix
  sampleGroups <- cutree(hclust(as.dist(1 - sampleCor), method = "ward.D2"), n_clust)  

  # Create an annotation data frame with cluster assignments
  anno_df <- data.frame(clust = as.character(sampleGroups), row.names = colnames(dds))  

  # Generate the heatmap plot using pheatmap
  pheatmap(as.matrix(sampleCor), 
          col = colorpanel(100, "#2b8cbe", "#7bccc4", "white"),
          clustering_distance_rows = as.dist(1 - sampleCor),  
          clustering_distance_cols = as.dist(1 - sampleCor),  
          clustering_method = "ward.D2",  
          annotation_col = anno_df,
          main = "Heatmap of sample-to-sample Correlation") 

}

plotHmapCorrelation(vsd, 2)

```

# Summary

```{r session info}
sessionInfo()
```

# Reference

1. https://chipster.csc.fi/manual/deseq2.html. n.d. https://chipster.csc.fi/manual/deseq2.html

2. Love, Simon Anders, and Wolfgang Huber, M. I. 2022, November 23. Analyzing RNA-seq data with DESeq2. Analyzing RNA-seq data with DESeq2. https://www.bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#howto-get-help-for-deseq2.



Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
